core/src/main/java/org/renjin/primitives/S3.java

```diff
@@ -259,6 +261,209 @@ public static SEXP tryDispatchFromPrimitive(Context context, Environment rho, Fu
+        evaluatedArgSuperClasses[i] = (String) evaledArgSupClass[i];
+      }
+
+      Object[] evalDist = superClassesAndDistance.values().toArray();
+      int[] evaluatedArgSuperClassesDistance = new int[evalDist.length];
+      for(int i = 0; i < evalDist.length; i++) {
+        evaluatedArgSuperClassesDistance[i] = (Integer) evalDist[i];
+      }
+
+      String[] allArgClasses = new String[evaluatedArgSuperClasses.length + 1];
+      int[] allArgClassesDistance = new int[evaluatedArgSuperClasses.length + 1];
+      allArgClasses[0] = evaluatedArg;
+      allArgClassesDistance[0] = 0;
+      for(int i = 1; i < evaluatedArgSuperClasses.length + 1; i++) {
+        allArgClasses[i] = evaluatedArgSuperClasses[i - 1];
+        allArgClassesDistance[i] = evaluatedArgSuperClassesDistance[i - 1];
+      }
+
+      if(signature == null) {
+        current++;
+        signature = new Object[allArgClasses.length][2];
+        for(int i = 0; i < allArgClasses.length; i++) {
+          signature[i][0] = allArgClasses[i];
+          signature[i][1] = allArgClassesDistance[i];
+        }
+        return generateAllPossibleSignatures(context, rho, args, signature, depth, current);
+      }
+
+      int rows = signature.length * allArgClasses.length;
+      int cols = depth + 1;
+      Object[][] newSig = new Object[rows][cols];
+
+      int lastSigIdx = 0; int lastSigMax = signature.length;
+      int currSigIdx = 0; int currSigMax = allArgClasses.length;
+      for(int row = 0; row < rows; row++) {
+        if (lastSigIdx == lastSigMax) {
+          lastSigIdx = 0;
+          currSigIdx++;
+        }
+        if (currSigIdx == currSigMax) {
+          currSigIdx = 0;
+        }
+
+        newSig[row][0] = signature[lastSigIdx][0] + "#" + allArgClasses[currSigIdx];
+
+        for (int col = 1; col < current + 1; col++) {
+          newSig[row][col] = signature[lastSigIdx][col];
+        }
+
+        newSig[row][current + 1] = allArgClassesDistance[currSigIdx];
+        lastSigIdx++;
+      }
+      current++;
+      return generateAllPossibleSignatures(context, rho, args, newSig, depth, current);
+    }
+
+    return signature;
+  }
+
+  public static String evaluateAndGetClass(Context context, PairList args, Environment rho, int argumentIndex) {
+    // To get the class of an argument we have to evaluate the argument first. If its an atomic, than it lacks class
+    // attribute. In this case we use Attributes.getClass() to get the class name.
+    String className;
+    SEXP evaluatedArg = context.evaluate(args.getElementAsSEXP(argumentIndex), rho);
+    if(evaluatedArg.getAttributes().hasClass()) {
+      className = evaluatedArg.getAttributes().getClassVector().getElementAsString(0);
+    } else {
+      className = Attributes.getClass(evaluatedArg).getElementAsString(0);
+    }
+    return className;
+  }
+
+  public static Map<String, Integer> getSuperClasses(Context context, ArrayList<String> allArgClasses, int signatureLength, int current) {
```

Comment:
allArgClasses as opposed to what? Do you have another list of arguments that does not include all arguments? 
"current" what?